hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java

```diff
@@ -82,39 +83,91 @@ public WanReplicationServiceImpl(Node node) {
     }
 
     @Override
-    public WanReplicationPublisher getWanReplicationPublisher(String name) {
+    public WanReplicationPublisherDelegate getWanReplicationPublishers(String name) {
         return getOrPutSynchronized(wanReplications, name, this, publisherDelegateConstructorFunction);
     }
 
-    private WanReplicationEndpoint[] createEndpoints(WanReplicationConfig wanReplicationConfig,
-                                                     Collection<? extends AbstractWanPublisherConfig> publisherConfigs) {
-        WanReplicationEndpoint[] targetEndpoints = new WanReplicationEndpoint[publisherConfigs.size()];
-        int count = 0;
-        for (AbstractWanPublisherConfig pc : publisherConfigs) {
-            final WanReplicationEndpoint target = getOrCreate((WanReplicationEndpoint) pc.getImplementation(),
-                    node.getConfigClassLoader(),
-                    pc.getClassName());
-            if (target == null) {
-                throw new InvalidConfigurationException("Either \'implementation\' or \'className\' "
-                        + "attribute need to be set in the WAN publisher configuration for publisher " + pc);
-            }
-            target.init(node, wanReplicationConfig, pc);
-            targetEndpoints[count++] = target;
+    private ConcurrentMap<String, WanReplicationPublisher> createPublishers(WanReplicationConfig wanConfig) {
+        List<CustomWanPublisherConfig> customPublisherConfigs = wanConfig.getCustomPublisherConfigs();
+        int publisherCount = customPublisherConfigs.size();
+
+        if (publisherCount == 0) {
+            return createConcurrentHashMap(1);
+        }
+
+        ConcurrentMap<String, WanReplicationPublisher> publishers = createConcurrentHashMap(publisherCount);
+        Map<String, AbstractWanPublisherConfig> publisherConfigs = createHashMap(publisherCount);
+
+        customPublisherConfigs.forEach(
+                publisherConfig -> {
+                    String publisherId = getPublisherIdOrGroupName(publisherConfig);
```

Comment:
`null` is a valid value for `publisherId`. This may lead NPE.

Reply:
Got it, will add some checks for this.

Reply:
seem this was not addressed?